Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-140.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: 9332e7fc59f463d5ca40a8d9ea0de81b433f46f4
Containerized:False
Number of Error(s): 8
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-15903/HPCC-Platform/thorlcr/master/thmastermain.cpp:1049:34: error: \xe2\x80\x98workunit\xe2\x80\x99 was not declared in this scope
/home/centos/smoketest/PR-15903/HPCC-Platform/thorlcr/master/thmastermain.cpp:1049:44: error: \xe2\x80\x98graphName\xe2\x80\x99 was not declared in this scope
make[2]: *** [thorlcr/master/CMakeFiles/thormaster_lcr.dir/build.make:118: thorlcr/master/CMakeFiles/thormaster_lcr.dir/thmastermain.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:16746: thorlcr/master/CMakeFiles/thormaster_lcr.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2022-03-16_12-39-39
time logs started
Prep time   :    15 sec (00:00:15)
Build time  :   340 sec (00:05:40)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   355 sec (00:05:55)
time logs end

[Build error:](https://gist.github.com/525eae5622de3f5fbcc37d754fab636d)
Build log [Build log](https://gist.github.com/HPCCSmoketest/525eae5622de3f5fbcc37d754fab636d#file-result-22-03-16-12-33-28-log)

--- a/thorlcr/master/thmastermain.cpp
+++ b/thorlcr/master/thmastermain.cpp
@@ -647,6 +647,8 @@ int main( int argc, const char *argv[]  )
 
     installDefaultFileHooks(globals);
     ILogMsgHandler *logHandler;
+    const char *workunit = nullptr;
+    const char *graphName = nullptr;
     try
     {
 #ifndef _CONTAINERIZED
@@ -824,7 +826,14 @@ int main( int argc, const char *argv[]  )
         if (l) { thorPath[l] = PATHSEPCHAR; thorPath[l+1] = '\0'; }
         globals->setProp("@thorPath", thorPath);
 
-#ifndef _CONTAINERIZED
+#ifdef _CONTAINERIZED
+        workunit = globals->queryProp("@workunit");
+        graphName = globals->queryProp("@graphName");
+        if (isEmptyString(workunit))
+            throw makeStringException(0, "missing --workunit");
+        if (isEmptyString(graphName))
+            throw makeStringException(0, "missing --graphName");
+#else
         const char * overrideBaseDirectory = globals->queryProp("@thorDataDirectory");
         const char * overrideReplicateDirectory = globals->queryProp("@thorReplicateDirectory");
         StringBuffer datadir;
@@ -870,6 +879,14 @@ int main( int argc, const char *argv[]  )
                 tempDirStr.append("temp");
             }
         }
+
+#ifdef _CONTAINERIZED
+        // multiple thor jobs can be running on same node, sharing same local disk for temp storage.
+        // make unique by adding wuid+graphName
+        addPathSepChar(tempDirStr).append(workunit);
+        addPathSepChar(tempDirStr).append(graphName);
+#endif
+        // NB: set into globals, serialized and used by worker processes.
         globals->setProp("@thorTempDirectory", tempDirStr);
         logDiskSpace(); // Log before temp space is cleared
         StringBuffer tempPrefix("thtmp");
@@ -918,15 +935,7 @@ int main( int argc, const char *argv[]  )
         kjServiceMpTag = allocateClusterMPTag();
 
         unsigned numWorkers = 0;
-        const char *workunit = nullptr;
-        const char *graphName = nullptr;
 #ifdef _CONTAINERIZED
-        workunit = globals->queryProp("@workunit");
-        graphName = globals->queryProp("@graphName");
-        if (isEmptyString(workunit))
-            throw makeStringException(0, "missing --workunit");
-        if (isEmptyString(graphName))
-            throw makeStringException(0, "missing --graphName");
         LogMsgJobId thorJobId = queryLogMsgManager()->addJobId(workunit);
         thorJob.setJobID(thorJobId);
         setDefaultJobId(thorJobId);

--- a/thorlcr/slave/thslavemain.cpp
+++ b/thorlcr/slave/thslavemain.cpp
@@ -487,11 +487,8 @@ int main( int argc, const char *argv[]  )
             }
 #endif
 
-            StringBuffer tempDirStr;
-            if (getConfigurationDirectory(globals->queryPropTree("Directories"),"spill","thor",globals->queryProp("@name"), tempDirStr))
-                globals->setProp("@thorTempDirectory", tempDirStr.str());
-            else
-                tempDirStr.append(globals->queryProp("@thorTempDirectory"));
+            // NB: master has set, and serialized in globals
+            StringBuffer tempDirStr(globals->queryProp("@thorTempDirectory"));
             addPathSepChar(tempDirStr).append(mySlaveNum);
 
             logDiskSpace(); // Log before temp space is cleared

